src/runtime_src/core/common/api/xrt_module.cpp

```
@@ -425,27 +437,72 @@ generate_key_string(const std::string& argument_name, xrt_core::patcher::buf_typ
   return argument_name + buf_string;
 }
 
+
+// Basic Itanium ABI type decoding. demangle_type() is referenced from ChatGPT response
 static std::string
-demangle(const std::string& mangled_name)
+demangle_type(const std::string& s, size_t& idx)
 {
-#ifdef _WIN32
-  char demangled_name[1024];
-  if (UnDecorateSymbolName(mangled_name.c_str(), demangled_name, sizeof(demangled_name), UNDNAME_COMPLETE))
-    return std::string(demangled_name);
-  else
-    throw std::runtime_error("Error demangling kernel signature");
-#else
-  int status = 0;
-  std::unique_ptr<char, decltype(&std::free)> demangled_name
-    (abi::__cxa_demangle(mangled_name.c_str(), nullptr, nullptr, &status), std::free);
+  if (idx >= s.size())
+    return "?";
 
-  if (status)
-    throw std::runtime_error("Error demangling kernel signature");
+  char c = s[idx++];
 
-  return demangled_name.get();
-#endif
+  // Append "*" for pointer types
+  if (c == 'P')
+    return demangle_type(s, idx) + "*";
+
+  auto it = demangle_type_map.find(c);
+  return (it != demangle_type_map.end()) ? it->second : "unknown";
+}
+
+// Parse mangled name in Itanium ABI style: _Z<length><name><types>
+// length : number of characters in the name string.
+// name   : kernel name in string
+// types  : kernel argument data type as below.
+//  'c' represents the arg is a char.
+//  'v' represents the arg is a void.
+//  'i' represents the arg is an int.
+//  'P' represents the arg is a pointer.
+//      Hence, "Pc" = char*, "Pv" = void*, "Pi" = int*
+// demangle() is referenced from ChatGPT response
+static std::string
+demangle(const std::string& mangled)
+{
+  //Check if mangled prefix "_Z" is present and length is greater than mangled_prefix_length
+  if (mangled.size() <= mangled_prefix_length || mangled.substr(0, mangled_prefix_length) != "_Z")
+    return "Not a mangled name";
```

We should throw instead of returning